### PR TITLE
Implementation of FragmentRequestMapper 

### PR DESCRIPTION
**Pull request by ssmusoke**
That allows requests to fragments to be overridden by implementations within local modules

### CODE REVIEW

File: api/src/main/java/org/openmrs/ui/framework/fragment/FragmentRequestMapper.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package org.openmrs.ui.framework.fragment;
+
+/**
+ * Maps a fragment request to an internal fragment name, e.g. you might map the request
+ * for the 'patient' fragment to 'clinicianPatientDashboard' if the user is logged in as a
```
Comment (djazayeri) [inline api/src/main/java/org/openmrs/ui/framework/fragment/FragmentRequestMapper.java]: Change the example in the javadoc to be a realistic one based on your use case.

File: api/src/main/java/org/openmrs/ui/framework/page/PageFactory.java [hunk 2]
```diff
@@ -43,6 +40,8 @@
 	
 	@Autowired(required=false)
 	List<PageRequestMapper> requestMappers;
+	@Autowired(required=false)
```
Comment (djazayeri) [inline api/src/main/java/org/openmrs/ui/framework/page/PageFactory.java]: Shouldn't this be in FragmentFactory, not PageFactory?

**Comment by djazayeri** [comment]
Unfortunately the FragmentFactory page got completely overwritten in this commit (maybe because of formatting), so I can't actually see what's happening there. Can you redo this PR in a way that makes the changes obvious/visible?

**Comment by ssmusoke** [comment]
Darius,

Sorry for the overwrite, let me update and resubmit the PR

Stephen

Stephen S. Musoke
Skype: ssmusoke
Twitter: @ssmusoke
+256 776 553 260
http://ssmusoke.com
*Our deepest fear is not that we are inadequate. Our deepest fear is that
we are powerful beyond measure ~Marianne Williamson*

**Comment by ssmusoke** [comment]
@djazayeri I am still struggling to remove the whitespace, but I found that adding w=1 to the url helps fix the whitespace issues, so is reviewing the PR via this link helpful https://github.com/openmrs/openmrs-module-uiframework/pull/37/files?w=1

File: api/src/main/java/org/openmrs/ui/framework/fragment/FragmentFactory.java [hunk 3]
```diff
@@ -1,576 +1,600 @@
...
+import org.openmrs.api.APIAuthenticationException;
+import org.openmrs.ui.framework.RequestValidationException;
+import org.openmrs.ui.framework.UiFrameworkException;
+import org.openmrs.ui.framework.UiFrameworkUtil;
+import org.openmrs.ui.framework.UiUtils;
+import org.openmrs.ui.framework.extension.ExtensionManager;
+import org.openmrs.ui.framework.formatter.FormatterService;
+import org.openmrs.ui.framework.fragment.action.FailureResult;
+import org.openmrs.ui.framework.interceptor.FragmentActionInterceptor;
+import org.openmrs.ui.framework.page.*;
```
Comment (djazayeri) [inline api/src/main/java/org/openmrs/ui/framework/fragment/FragmentFactory.java]: We don't like wildcard imports in OpenMRS (except in test classes). Can you please change this IntelliJ setting?

File: api/src/main/java/org/openmrs/ui/framework/fragment/FragmentFactory.java [hunk 4]
```diff
@@ -1,576 +1,600 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.MessageSource;
+import org.springframework.core.convert.ConversionService;
+
+import javax.servlet.ServletContext;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpSession;
+import java.lang.reflect.Method;
+import java.util.*;
```
Comment (djazayeri) [inline api/src/main/java/org/openmrs/ui/framework/fragment/FragmentFactory.java]: Same comment: avoid wildcard imports.

File: api/src/main/java/org/openmrs/ui/framework/fragment/FragmentFactory.java [hunk 5]
```diff
@@ -1,576 +1,600 @@
...
+		FragmentRequest request = new FragmentRequest(providerName, fragmentName);
+		return getController(request);
+	}
+	
+	/**
+	 * Allow modules to override fragment handling via {@link FragmentRequestMapper}s.
+	 * Sets this internal fragment provider and fragmentId on request
+	 * @param request
+	 */
+	private FragmentRequest mapInternalFragmentId(FragmentRequest request) {
```
Comment (djazayeri) [inline api/src/main/java/org/openmrs/ui/framework/fragment/FragmentFactory.java]: I assume that you copy-paste-edited this name from PageFactory, but I don't actually like it. (I don't know what I was thinking years ago...). Can you rename this method to "maybeOverrideFragmentProviderAndId"?

**Comment by djazayeri** [comment]
I looked at the ?w=1 url, and this is basically good, except for the few comments I made.

But please find a way to do the PR such that it doesn't overwrite all the lines in FragmentFactory.

File: api/src/main/java/org/openmrs/ui/framework/fragment/FragmentFactory.java (hunk 5 above)
Comment (ssmusoke) [inline api/src/main/java/org/openmrs/ui/framework/fragment/FragmentFactory.java]: @djazayeri Does it also make sense to update the method call PageFactory method too in order to maintain consistency? 

I am suggesting use of overrideMappedProviderNameAndFragmentId as the FragmentRequest class has been updated to always use a mapped provider name and fragmentid which is being overridden

Comment (djazayeri) [inline api/src/main/java/org/openmrs/ui/framework/fragment/FragmentFactory.java]: Yes it makes sense to change the other one for consistency.

Actually a better name is just "mapFragmentProviderAndId".

**Comment by ssmusoke** [comment]
Closing this PR to start over again due to line ending issues with the FragmentFactory class

**Comment by ssmusoke** [comment]
Closed fixed line issue formatting of the FragmentFactory class in https://github.com/openmrs/openmrs-module-uiframework/pull/38

**Pull request closed without merging**